apps/emqx_durable_storage/src/emqx_ds_storage_bitfield_lts.erl

```diff
@@ -412,11 +410,10 @@ make_key(#s{keymappers = KeyMappers, trie = Trie}, #message{timestamp = Timestam
 ]) ->
     binary().
 make_key(KeyMapper, TopicIndex, Timestamp, Varying) ->
-    UniqueInteger = erlang:unique_integer([monotonic, positive]),
```

Comment:
Since we want total order of messages per stream, this PR just enforces total order per shard through allocating [unique pseudo-realtime timestamps](https://github.com/emqx/emqx/blob/c199c0b1613b7de18f760eb1e1cf4a1efc9e11d0/apps/emqx_durable_storage/src/emqx_ds_replication_layer.erl#L491) to messages (admittedly, in a quite ugly way). Since we don't want timestamps to occasionally jump backwards during leader election it's important to have that as part of Raft log / state, and uniqueness (i.e. strict monotonicity) is trivial to add on top of that.

With that we don't need this extra uniqueness, which furthermore is not really guaranteed to be unique across shard's replica set.

Reply:
I'm thinking about if that could also be a problem, now from the opposite side.  :sweat_smile: 

When [moving session data from mnesia to DS](https://github.com/emqx/emqx/pull/12675), I took advantage of the fact that the message timestamps are respected by the storage layer to upsert keys.  Now that `emqx_ds_replication_layer:assign_timestamps` enforces unique timestamps, it'll no longer be possible to upsert a key if one wants to do that in a particular DB.  :thinking:

Reply:
> Now that `emqx_ds_replication_layer:assign_timestamps` enforces unique timestamps, it'll no longer be possible

Conceptually, this needs to be enforced only for messages that needs to be replayed in a strict order. If we need to accommodate for another storage semantics, nothing should stop us from expanding API here, e.g. adding flags or properties to _store batch_ command.